### PR TITLE
osd: Implement Context based completion for mon cmd to set a config option

### PR DESCRIPTION
**Pull request by sseshasa**
The method, OSD::mon_cmd_set_config() currently sets a config option
related to mClock during OSD boot-up. The method used to wait on a
condition variable until the mon ack'ed the command. This was generally
not a problem.

But there could be scenarios where monitor could be slow to respond, or
due to a flaky network, response could be delayed. The OSD could therefore
be blocked from booting-up. To avoid this, the conditional wait can be
replaced with an async Context completion.

Moreover, persisting this in the monitor store is not very critical. An
existing fallback mechanism stores this value in the in-memory "values"
map of the config subsystem. This can be read by the OSD at any point
during its operation.

The issue of the OSDs being blocked from booting-up properly was
observed when running tests with failure injections during OSD boot-up.

Following are the changes:

The changes to mon_cmd_set_config() are generic and any osd specific
option may be set in the config monitor store using this method.

- Implement Context based completion tracking of the mon command using
  MonCmdSetConfigOnFinish which is derived from the base Context class.
  In case of failures, the finish() method is overriden to save the config
  option in the config subsystem's "values" map at CONF_DEFAULT level
  using set_val_default() method. This allows users to modify this at a
  later point using "ceph config set" cli command.

- Additionally, if requested, finish() also checks if the config option
  needs to be applied on each op shard and calls update_scheduler_config()
  on each shard. This is required for some config options related to the
  mClock scheduler.

Fixes: https://tracker.ceph.com/issues/57040
Signed-off-by: Sridhar Seshasayee <sseshasa@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by sseshasa** [comment]
**Teuthology Tests**
https://pulpito.ceph.com/sseshasa-2022-08-03_18:56:32-smoke:basic-wip-sseshasa-testing-mon-nack-distro-default-smithi/
https://pulpito.ceph.com/sseshasa-2022-08-04_08:53:37-rbd:nbd-wip-sseshasa-testing-mon-nack-distro-default-smithi/

The above are thrash tests with msgr failures. The logs from the tests were analyzed and it was confirmed that the
OSDs now boots up without waiting, and sets the shard specific IOPS capacity when the Context is completed.

**Comment by sseshasa** [comment]
jenkins test make check

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -9868,7 +9859,41 @@ bool OSD::maybe_override_options_for_qos()
...
+  std::string key;
+  std::string val;
+  bool update_shard;
+public:
+  explicit MonCmdSetConfigOnFinish(
+    OSD *o,
+    CephContext *cct,
+    const std::string &k,
+    const std::string &v,
+    const bool &s)
```
Comment (athanatos) [inline src/osd/OSD.cc:9878]: No need to pass a bool by ref.

**Review by athanatos — APPROVED** [review]
LGTM

File: src/osd/OSD.cc (hunk 1 above)
Comment (sseshasa) [reply src/osd/OSD.cc]: Fixed.

**Comment by sseshasa** [comment]
jenkins test make check

**Comment by sseshasa** [comment]
jenkins test api

**Comment by sseshasa** [comment]
jenkins test windows

**Comment by sseshasa** [comment]
jenkins render docs

**Comment by sseshasa** [comment]
jenkins test docs

**Comment by sseshasa** [comment]
jenkins retest this please

**Comment by sseshasa** [comment]
jenkins test windows

**Comment by sseshasa** [comment]
jenkins test make check

**Comment by sseshasa** [comment]
jenkins test windows

**Comment by sseshasa** [comment]
jenkins test windows

**Comment by sseshasa** [comment]
jenkins test windows

**Comment by sseshasa** [comment]
jenkins test windows

**Comment by sseshasa** [comment]
**Teuthology Test Result**
http://pulpito.front.sepia.ceph.com/?branch=wip-yuri5-testing-2022-08-18-0812
Summary: 294/319 Passed, 22 Failed, 3 dead jobs

**Unrelated Failures**
The following are the unrelated failures for which trackers already exist:

1. https://tracker.ceph.com/issues/57368 - The CustomResourceDefinition "installations.operator.tigera.io" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
2. https://tracker.ceph.com/issues/57122 - test failure: rados:singleton-nomsgr librados_hello_world - Ceph - RADOS
3. https://tracker.ceph.com/issues/55853 - cls_rgw test failures
4. https://tracker.ceph.com/issues/57165 - expected valgrind issues and found none 
5. https://tracker.ceph.com/issues/57270 - stderr E: The repository 'https://download.ceph.com/debian-octopus jammy Release' does not have a Release file
6. https://tracker.ceph.com/issues/52321 - qa/tasks/rook times out: 'check osd count' reached maximum tries (90) after waiting for 900 seconds - Ceph - Orchestrator
7. https://tracker.ceph.com/issues/56850 - src/mon/PaxosService.cc: 193: FAILED ceph_assert(have_pending)

**Comment by sseshasa** [comment]
jenkins test windows

**Pull request merged**